stop: run closers in reverse order

RaduBerinde opened the pull request:
#### stop: run closers in reverse order

We also fix the order of a few `AddCloser()` calls.

In particular, we move the closer for the deletion of the temp store
directory where it belongs (next to the code that creates it). We
modify the code a bit to make sure we can unlock the directory before
trying to clean it up.

Fixes: #142347

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143431)
<!-- Reviewable:end -->

tbg reviewed (approved):
LGTM
This is the kind of change that could have some unexpected fallout, so let's keep our eyes peeled.

RaduBerinde commented:
I agree. I could hold off on merging until the release-25.2 branch fork.

RaduBerinde committed:
stop: run closers in reverse order

We also fix the order of a few `AddCloser()` calls.

In particular, we move the closer for the deletion of the temp store
directory where it belongs (next to the code that creates it). We
modify the code a bit to make sure we can unlock the directory before
trying to clean it up.

Fixes: #142347

RaduBerinde commented:
bors r+
